### PR TITLE
3D Video Playback Timeline

### PR DESCRIPTION
**Pull request by benptc**
Adds a quite complex timeline system that can be opened thru the View Menu. It will load all recorded 3D videos onto the timeline, organized by day and by track (each unique recording device used on that day gets a track). Data can be zoomed, scrolled, played/paused, and playback speed can be adjusted.

After building this as a monolithic Timeline class that was suffocating under its own weight I've refactored it into a MVC design with additional components. Everything starts with the VideoPlaybackCoordinator, which creates the TimelineController and connects its video data output to the CameraVisCoordinator to render the point-clouds. The TimelineView renders all the timeline UI according to the TimelineModel state. TimelineModel contains a TimelineDatabase, which is populated with video data from the VideoSources.

The timeline has been designed for a future in which more data can be added to it as additional tracks. For example, a pose data track or IoT data track could be added, and the timeline will output the time series data along with the videos.

It's far from perfect but I'm running out of time to work on this for now, so trying to get a PR in!

_Note: the one existing area I slightly refactored is the CameraVis, where I pulled out most of the contents of `connectWsToTexture` and `connectWsToMatrix` to their own functions so that I could call them from the playback coordinator._

Screen Shot 2022-05-04 at 8 35 17 PM

### CODE REVIEW

**Comment by benptc** [comment]
Sorry for the huge PR 👀 this grew way bigger than I was expecting and frankly I probably shouldn't have spent this much time on this 😐 hopefully it proves useful sometime in the distant future, but for now I need to work on other things!

**Review by hobinjk-ptc — APPROVED** [review]
This is super cool! Minor nitpicks inline

File: content_scripts/desktopAdapter.js [hunk 1]
```diff
@@ -392,6 +395,13 @@ window.DEBUG_DISABLE_DROPDOWNS = false;
      */
     function addSocketListeners() {
 
+        realityEditor.network.realtime.addDesktopSocketMessageListener('disconnect', function(_msgContent) {
```
[the remark this answers is not in the document]
Comment (benptc) [reply content_scripts/desktopAdapter.js:398]: Oops, I'm gonna get rid of that listener for now, I had the idea of showing some visual feedback if the server disconnects but never followed through with the implementation.

File: content_scripts/desktopCamera.js [hunk 2]
```diff
@@ -146,9 +146,9 @@ createNameSpace('realityEditor.device.desktopCamera');
             pointerPosition.x = e.clientX;
             pointerPosition.y = e.clientY;
 
-            if (interactionCursor && interactionCursor.style.display !== 'none') {
-                interactionCursor.style.left = (pointerPosition.x - interactionCursor.getClientRects()[0].width/2) + 'px';
-                interactionCursor.style.top = (pointerPosition.y - interactionCursor.getClientRects()[0].height/2) + 'px';
+            if (interactionCursor && interactionCursor.getClientRects() && interactionCursor.getClientRects()[0] && interactionCursor.style.display !== 'none') {
```
[the remark this answers is not in the document]
Comment (benptc) [reply content_scripts/desktopCamera.js]: Good feedback! I created a helper function to get this in a consistent/efficient way.

**Pull request merged**